### ISSUE TITLE
Propose "tc-style" and possibly viewtemplate style setter

### ISSUE THREAD

**Issue by twMat** [Jermolene/TiddlyWiki5#1913]
It is very cool that you can now [Apply custom styles by tag](http://tiddlywiki.com/#How%20to%20apply%20custom%20styles%20by%20tag), i.e so that indviidual tiddlers can be styled by defining a CSS class with the name `tc-tagged-<TagName>`.

However, it demands that the user fiddles with css and stylesheets. 

Here follows an idea for an implementation that would make it super easy to style individual tiddlers. 

First a prototype, based on the current tc-tagged feature so to illustrate the idea and why it would be much better if we instead could use "tc-fields" (perhaps titled tc-style). I.e this works :

```
<$set name="this" value=<<currentTiddler>>>
<$list filter="[all[current]tags[]]" variable="item">
<$list filter="[<item>] +[splitbefore[:]]" variable="property">
<$list filter="[<item>] +[removeprefix<property>]" variable="value">
<style>
.tc-tagged-stylethis { <<property>><<value>>; }
</style>
</$list>
</$list>
</$list>
</$set>
```

The user must tag the tiddler on the form `property:value`, e.g `background-color:yellow` - and he needs to tag it "stylethis". Not very elegant - but this can be much improved by introducing `tc-style` instead of `tc-tags`. The user _could_ then manually define `tc-style-<Field Name>` but...

**...a modified version of the above to get the values from the current tiddler _fields_ would be superior, i.e so a user can create _fields_ like `background-color:yellow`**... and the field drop down list can be predefined to show the standard CSS styling properties. A user shouldn't have to keep these things in mind or search for them.

This would make styling individual tiddlers super easy.

**Comment by pmario** [comment]
interesting approach. .. but where do you insert your prototype template?

did you test it? does it work?

**Comment by twMat** [comment]
Note: I mistakenly referred to "attributes" when I meant "properties" in the above. Anyway...

Here's a minimal [demo](http://selfstyle.tiddlyspot.com/). Does this answer your question, (or else what do you mean with prototype template?)

**Comment by tobibeer** [comment]
Polluting the tag-space this way seems like a bad idea. Also, the tagging approach is helpful when styling tiddlers pertaining to categories.

If at all, I'd rather a "css" field. And then split proper css...

```
<$list filter="[all[current]has[css]]">
<$list filter="[all[current]get[css]] + [splitbefore[;]]" variable="style">
<$list filter="[<style>] +[splitbefore[:]]" variable="property">
<$list filter="[<style>] +[removeprefix<property>]" variable="value">
<style>
.tc-tagged-stylethis { <<property>><<value>>; }
</style>
</$list>
</$list>
</$list>
</$list>
```

**Comment by twMat** [comment]
Hm,  my *actual* proposal was a bit unclear; one of my intended conclusive points was to get away from tags all together (for a "behind-the-scene" matter like styling) and *only* use fields.

I imagine using one field per property that one wants to manipulate. It would allow the field dropdown list to be pre-populated and the values to behave "accordingly" (e.g a field titled "color" would get the color picker).

(BTW, your idea seem to only handle one CSS property or am I missing some recursion or something?)

**Comment by tobibeer** [comment]
Yes, just one css field to declare all the styles for one element, namely the tiddler.

I also think having one field per css property would likewise pollute the namespace for fields, especially with the new fieldname popup. It would overly mingle actual content fields with these "style fields". Besides, how would you know that a field is a style field? Also, this would prevent you from using css properties as a field name for other content, because it would be interpreted as if holding a proper css value. Using just one style field, it's pretty clear what to dissect.

All in all, I do not see much benefit from styling nothing but the tiddler element this way. Using stylesheet tiddlers clearly is the way to go, focusing on one method, rather than providing ever more ways of doing the same thing.

**Comment by tobibeer** [comment]
Consider [this solution posted by Eric](https://groups.google.com/d/msg/tiddlywiki/0iMvwZkyKrA/zOnNoBU5DAAJ) instead, where custom styles per tiddler are stored in and pulled from a custom tiddler field:

> If you want to apply custom styles **without defining classnames**, then this ViewTemplate is what you want:

    \define frame-classes()
    tc-tiddler-frame tc-tiddler-view-frame $(missingTiddlerClass)$ $(shadowTiddlerClass)$ $(systemTiddlerClass)$ $(tiddlerTagClasses)$
    \end
    <$set name="storyTiddler" value=<<currentTiddler>>><$set name="tiddlerInfoState" value=<<qualify "$:/state/popup/tiddler-info">>><$tiddler tiddler=<<currentTiddler>>>
    <div class=<<frame-classes>> style={{!!styles}}><$list filter="[all[shadows+tiddlers]tag[$:/tags/ViewTemplate]!has[draft.of]]" variable="listItem"><$transclude tiddler=<<listItem>>/></$list>
    </div>
    </$tiddler></$set></$set>

> Note how this is very similar to using a custom class, but the syntax is even simpler.  Just add `"style={{!!styles}}"` to the attributes of the tiddler frame.  Then create a custom **"styles"** field in a tiddler and enter your CSS attribute definitions there.

**Comment by twMat** [comment]
@tobibeer wrote Sept 16
>I also think having one field per css property would likewise pollute the namespace for fields, especially with the new fieldname popup.

My point with "one field per css property" is so that they can contain //predefined// values, like the color field already does - so that the user does not have to know/memorize CSS property names or property values. (IMO a user should only have to "learn TW" not "learn CSS" to use TW.) The field name space is already over-populated but this is a UI matter, i.e we can design it better, perhaps a two level list. 

Regarding Erics solution (ref. post above this one) - I like much it better than being forced to tag a tiddler, or perhaps in //addition// to tagging, it but it suffers the "flaw" I just pointed out about learning. Ideally, I'd see both something along my OP and Erics, with the latter for a "quick'n dirty"... or perhaps it's just "quick" ;-)

**Comment by tobibeer** [comment]
> IMO a user should only have to "learn TW" not "learn CSS" to use TW.

I don't understand. Do you want to invent a whole new dictionary to translate well known css properties and markup to something else? How are there other values permissible other than those define by the css standard?

Styles have virtually nothing to do with the tiddler and its content... so you mingle content fields and style fields. Not something I would do. That is what css and classes for. If you must store css directly at a tiddler, at least store it all in one field.

Even if what you really wanted was a "tiddler style editor", there would be no need to store each property in its own field.

> The field name space is already over-populated but this is a UI matter, i.e we can design it better, perhaps a two level list and the list could also be a kind of search result based on what you begin to type.

Styles are styles and content is content. We can mix those in wikitext already, but we should not cater for "css fields", imho. All this would do is over-complicate the core with little to no gain.

>  it suffers the "flaw" I just pointed out about learning. 

There is no getting around learning that. If you use one field per style: you still need to know what fields you can use and what syntax their properties require. So, you're back to square 1.

**Comment by twMat** [comment]
>so you mingle content fields and style fields. 
>...
>If you use one field per style: you still need to know what fields you can use and what syntax their properties require. 

Seems to work with the **color** field, no... ?

The problem is that the "well known CSS properties" are not well known unless you're well familiar with CSS. It is unfortunate to force users to study CSS merely to be able to style their content in TW. It's a threshold to use TW to begin with.

I possibly have a naive view on what fields are - I see them as basic titled containers of content. What the content (the value) is used for is another matter.

**Comment by tobibeer** [comment]
> Seems to work with the color field, no... ?

Yes, because that was considered sort of a top-level property, so tag-pills (and accompanying icons) would be customizable, per tiddler. There can sure be more of those, and I would personally not want any of them in the form of fields.

>  It is unfortunate to force users to study CSS merely to be able to style their content in TW. It's a threshold to use TW to begin with.

This has virtually nothing to do with how these style properties are stored. Instead, it is about providing some user interface that makes styling of "certain things" easier.

By "simply" using a myriad of fields to store styles, nothing has been simplified, neither the learning experience, nor the complexity of the matter. Instead, you have increased complexity by further intermingling two very different aspects of a document: content and presentation.

**Comment by tobibeer** [comment]
By the way @twMat, your [ViewFields](https://groups.google.com/forum/?fromgroups=#!topic/tiddlywiki/0UqjoSDw7do) is a perfect example for why not to use fields... at least, it is for me, because... eventually, TiddlyWiki is not about constantly, manually tweaking styles ...it's about content.

**Comment by Jermolene** [comment]
I'm not a huge fan of slapping CSS rules onto individual tiddlers because of the way that it buries presentation information within tiddlers. Having said that, I'd like to make these sorts of enhancements as easy as possible to accomplish via a plugin.

I'd suggest that the "color" field is slightly different because it is semantically simpler: it's a standard convention for associating a colour with a tiddler. The field itself doesn't say anything about how it's used.

**Comment by twMat** [comment]
The discussion is definitely drifting here (much because of me of course) but I think *fields* are waaay to general to limit them to only deal with *certain* type of content. Surely, if we were to limit the content type, it would make more sense to first define *field types* (categories) to match the content types. BTW, there seem already to be implicit field types; the types we're discussing right now are those things seen at tiddler bottom in edit mode. But we treat all of the *title, tags and text* fields differently and there are meta data fields that are treated in yet other ways. To me it makes sense to have e.g "style fields", "content fields", "state fields" etc and to set behaviour for those types lobally (and potentially tiddler-locally overridden). 

>TiddlyWiki is not about constantly, manually tweaking styles ...it's about content.

I don't mean to patronize but; Style and form and content are *all* parts in conveying a message. **This** is a different message than *this* and if it appears in a standard tiddler or a red one definitely says different things. Actually, a tiddler shouldn't even have to be a "post-it note" at all, imo.

**Comment by tobibeer** [comment]
> The discussion is definitely drifting here (much because of me of course) but I think fields are waaay to general to limit them to only deal with certain type of content.

I could go with that. Perhaps we can introduce some generic way that would handle "system-like" fields by...

1. taking them out of the default edit-mode template 
2. grouping and placing them into a foldable section

...e.g. all prefixed `css-foo`.

**Comment by twMat** [comment]
@Jermolene 
>I'm not a huge fan of slapping CSS rules onto individual tiddlers because of the way that it buries presentation information within tiddlers. 

I just made a [post](https://groups.google.com/forum/#!topic/tiddlywiki/EfYp47j-I_A) with a potential approach for this problem. Would love to hear your opinion there on it.

>Having said that, I'd like to make these sorts of enhancements as easy as possible to accomplish via a plugin.

Any thoughts on how to enable a plugin to add tiddler specific styles without overwriting core tiddlers (and without tagging)?

**Comment by Jermolene** [comment]
> Any thoughts on how to enable a plugin to add tiddler specific styles without overwriting core tiddlers (and without tagging)?

That's a goal we haven't achieved yet, and is why I am interested in these discussions.

**Comment by twMat** [comment]
@tobibeer 
>Perhaps we can introduce some generic way that would handle "system-like" fields by...
>
> *  taking them out of the default edit-mode template

I'm thinking there, in the one and same tiddler, could be "fields that contain meta-data about other fields". Perhaps the field *foo* has meta-data fields *foo_visibility-in-viewmode*, *foo_position-in-viewmode* etc. Even *foo_belongs-to-group_bar*.

Fields suffixed with those "meta suffixes" are not openly displayed (just like some of the other core fields are not displayed) so they dont clutter the tiddler.

Or meta-data about a field is stored outside of the tiddler, in a central dictionary type system tiddler listing. *<tiddler:field>:<field-meta-data>. Whenever a user creates a (non standard) field, an entry and some default field-meta-data is created and stored in this central field dictionary.

> *   grouping and placing them into a foldable section

I get very positive ideas when I think of what you might mean here but please elaborate :-)

**Comment by tobibeer** [comment]
> Perhaps the field foo has meta-data fields

That's precisely the kind of "mess" we want to avoid. You simply do not want to define that for each and every tiddler. Makes sense?

So, as [elaborated here](https://groups.google.com/d/msg/tiddlywiki/0UqjoSDw7do/N3NYig9lDAAJ), I think the way to go are global "field config tiddlers" that specify how to handle as well as view or edit fields.

**Comment by twMat** [comment]
>>Perhaps the field foo has meta-data fields
>
>That's precisely the kind of "mess" we want to avoid. You simply do not want to define that for each and every tiddler. Makes sense?

If I understand you then I think it doesn't quite make 100% sense, but close ;-)  The thing is fields are not uniquely titled and it is *not* evident that I want all fields titled "special_secret" to be treated the same way. However, maybe there could be a way to locally override global field settings so that the global values only serve as default. Now //those// should then probably be stored locally...

Now, how do we tie this back to the OP which is about how to style individual tiddlers without using tags?

**Comment by tobibeer** [comment]
> Now, how do we tie this back to the OP which is about how to style individual tiddlers without using tags?

I think my answer on that is pretty clear. :-)

**Comment by pmario** [comment]
>Any thoughts on how to enable a plugin to add tiddler specific styles without overwriting core tiddlers (and without tagging)?

I think tagging is an easy way to handle "per tiddler" styling and it is part of the core already.

 - At some point users need to learn basic CSS rules, if they want to change TW styling! 
 - At some point users need to start changing the existing content, if they want to have more control over the layout. .. because otherwise we will infinitely bloat the core, if we try to make everything configurable.

What if I want to write "hintergrund:gelb" instead of "background-color:yellow". ... It would be possible to create a UI that way. ... But it doesn't make sense!

IMO adding functionality as the OP describes, adds more complexity, than what we can win with it.

**Comment by pmario** [comment]
@twMat .. imo solved since 5.1.16 see: https://tiddlywiki.com/#How%20to%20apply%20custom%20styles:%5B%5BHow%20to%20apply%20custom%20styles%5D%5D%20%5B%5BCustom%20styles%20by%20user-class%5D%5D

**Comment by twMat** [comment]
@pmario yes, thanks to you and much appreciated!
